virtualizarr/manifests/manifest.py

```diff
@@ -56,25 +62,131 @@ class ChunkManifest(BaseModel):
+
+    def __init__(self, entries: dict) -> None:
+        """
+        Create a ChunkManifest from a dictionary mapping zarr chunk keys to byte ranges.
+
+        Parameters
+        ----------
+        entries: dict
+            Chunk keys and byte range information, as a dictionary of the form
+
+                {
+                    "0.0.0": {"path": "s3://bucket/foo.nc", "offset": 100, "length": 100},
+                    "0.0.1": {"path": "s3://bucket/foo.nc", "offset": 200, "length": 100},
+                    "0.1.0": {"path": "s3://bucket/foo.nc", "offset": 300, "length": 100},
+                    "0.1.1": {"path": "s3://bucket/foo.nc", "offset": 400, "length": 100},
+                }
+        """
 
-    entries: Mapping[ChunkKey, ChunkEntry]
-    # shape_chunk_grid: Tuple[int, ...]  # TODO do we need this for anything?
+        # TODO do some input validation here first?
+        validate_chunk_keys(entries.keys())
+
+        # TODO should we actually optionally pass chunk grid shape in,
+        # in case there are not enough chunks to give correct idea of full shape?
+        shape = get_chunk_grid_shape(entries.keys())
+
+        # Initializing to empty implies that entries with path='' are treated as missing chunks
+        paths = np.empty(shape=shape, dtype=np.dtypes.StringDType())
+        offsets = np.empty(shape=shape, dtype=np.dtype("int32"))
+        lengths = np.empty(shape=shape, dtype=np.dtype("int32"))
+
+        # populate the arrays
+        for key, entry in entries.items():
+            try:
+                path, offset, length = entry.values()
+                entry = ChunkEntry(path=path, offset=offset, length=length)
+            except (ValueError, TypeError) as e:
+                msg = (
+                    "Each chunk entry must be of the form dict(path=<str>, offset=<int>, length=<int>), "
+                    f"but got {entry}"
+                )
+                raise ValueError(msg) from e
+
+            split_key = split(key)
+            paths[split_key] = entry.path
+            offsets[split_key] = entry.offset
+            lengths[split_key] = entry.length
+
+        self._paths = paths
+        self._offsets = offsets
+        self._lengths = lengths
 
-    @field_validator("entries")
     @classmethod
-    def validate_chunks(cls, entries: Any) -> Mapping[ChunkKey, ChunkEntry]:
-        validate_chunk_keys(list(entries.keys()))
+    def from_arrays(
+        cls,
+        paths: np.ndarray[Any, np.dtype[np.dtypes.StringDType]],
+        offsets: np.ndarray[Any, np.dtype[np.int32]],
+        lengths: np.ndarray[Any, np.dtype[np.int32]],
+    ) -> "ChunkManifest":
+        """
+        Create manifest directly from numpy arrays containing the path and byte range information.
+
+        Useful if you want to avoid the memory overhead of creating an intermediate dictionary first,
+        as these 3 arrays are what will be used internally to store the references.
+
```

Comment:
@ayushnag @sharkinsspatial you might want to try building numpy arrays of references and passing them to this new constructor instead to keep memory usage down.